### ISSUE TITLE
IRI connection retry attempts before error / quit

### ISSUE THREAD

**Issue by rosstimson** [SemkoDev/nelson.cli#14]
Working on getting Nelson working alongside IRI using Docker as Docker seemed to be the preferred route to take for deployments during early discussions in Slack.

I am currently experimenting with a [Docker Compose file ](https://gist.github.com/rosstimson/bc0e4bc881e2c333b5bcfdbade720d3d) for setting up the entire stack with a single command  however, Nelson throws and error and quits as there is a slight race condition and it tries to connect to IRI before the IRI container is ready.  I can workaround this by simply running `docker-compose start nelson` slightly after the initial `docker-compose up` but this isn't very user friendly.  It would be great if Nelson could simply try again a couple of times before erroring out.  Is that a possibility?

If you would prefer not to implement this there are a few possible workarounds:

## Option 1

Add a healthcheck to the IRI service, in Docker Compose you can do something like:

```
healthcheck:
  test: ["CMD", "curl", "-f", "http://localhost"]
  interval: 1m30s
  timeout: 10s
  retries: 3
```

The 'official' IRI Docker images does not have `curl` or anything else suitable to carry out a check so we'd need to create a custom image which ideally I'd like to avoid.  Also, this is a Docker Compose thing and would not help improve matters if you were deploying containers via some other method.

## Option 2

Amend the Nelson Dockerfile to change it to start via a small shell script that wraps the whole thing in a loop, something like:

```
while ! nc -z iri 14265; do sleep 3; done
```

I think option 2 is probably the better of the 2 here if you do not wish to make the code handle this.  I'd love to hear your thoughts.  This sort of thing is a common issue with Docker deployments, some further info and ideas can be seen here:  https://docs.docker.com/compose/startup-order/

**Comment by rosstimson** [comment]
Apologies, I completely forgot to post an example error about the issue I spoke about here.

```
nelson_1      | Unhandled Rejection at: Promise Promise {
nelson_1      |   <rejected> Error: Invalid Response: Error: connect ECONNREFUSED 172.19.0.2:14265
nelson_1      |     at Object.exports._errnoException (util.js:1022:11)
nelson_1      |     at exports._exceptionWithHostPort (util.js:1045:20)
nelson_1      |     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1087:14)
nelson_1      |     at Object.invalidResponse (/usr/local/lib/node_modules/nelson.cli/node_modules/iota.lib.js/lib/errors/requestErrors.js:5:12)
nelson_1      |     at makeRequest.prepareResult (/usr/local/lib/node_modules/nelson.cli/node_modules/iota.lib.js/lib/utils/makeRequest.js:160:24)
nelson_1      |     at request.onreadystatechange (/usr/local/lib/node_modules/nelson.cli/node_modules/iota.lib.js/lib/utils/makeRequest.js:62:25)
nelson_1      |     at dispatchEvent (/usr/local/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:591:25)
nelson_1      |     at setState (/usr/local/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:610:14)
nelson_1      |     at handleError (/usr/local/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:532:5)
nelson_1      |     at ClientRequest.errorHandler (/usr/local/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:459:14)
nelson_1      |     at emitOne (events.js:96:13)
nelson_1      |     at ClientRequest.emit (events.js:188:7)
nelson_1      |     at Socket.socketErrorListener (_http_client.js:310:9) } reason: Error: Invalid Response: Error: connect ECONNREFUSED 172.19.0.2:14265
nelson_1      |     at Object.exports._errnoException (util.js:1022:11)
nelson_1      |     at exports._exceptionWithHostPort (util.js:1045:20)
nelson_1      |     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1087:14)
nelson_1      |     at Object.invalidResponse (/usr/local/lib/node_modules/nelson.cli/node_modules/iota.lib.js/lib/errors/requestErrors.js:5:12)
nelson_1      |     at makeRequest.prepareResult (/usr/local/lib/node_modules/nelson.cli/node_modules/iota.lib.js/lib/utils/makeRequest.js:160:24)
nelson_1      |     at request.onreadystatechange (/usr/local/lib/node_modules/nelson.cli/node_modules/iota.lib.js/lib/utils/makeRequest.js:62:25)
nelson_1      |     at dispatchEvent (/usr/local/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:591:25)
nelson_1      |     at setState (/usr/local/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:610:14)
nelson_1      |     at handleError (/usr/local/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:532:5)
nelson_1      |     at ClientRequest.errorHandler (/usr/local/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:459:14)
nelson_1      |     at emitOne (events.js:96:13)
nelson_1      |     at ClientRequest.emit (events.js:188:7)
nelson_1      |     at Socket.socketErrorListener (_http_client.js:310:9)
nelsoncli_nelson_1 exited with code 0
```

**Comment by H3n-42** [comment]
Option 2 looks like a "hack" to me.

in my opinion the better way is to not error out nelson. It should just try to connect to the api, when it doesnt work, wait a few seconds and try again?

This way nelson and iri would work independely, and it doesnt breake when iri is not available for a few seconds

**Comment by rosstimson** [comment]
@H3npi Both options are pretty hacky, they are just necessary workarounds if the connection retry is not going to be implemented within Nelson itself.

**Comment by ghost** [comment]
And what about a ECONNRESET? After install I only get this with a locally installed IRI 1.4.1.2 up and running and no firewall:

```
Unhandled Rejection at: Promise Promise {
  <rejected> Error: Invalid Response: Error: read ECONNRESET
```

**Comment by jinnerbichler** [comment]
An intermediate solution can be found here: https://github.com/jinnerbichler/iota-node-docker
I am using a separate service to wait for the IRI node to start up.

**Comment by ghost** [comment]
I ran it by hand after IRI was full loaded.

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   597  100   572  100    25    572     25  0:00:01  0:00:01 --:--:--   548
{
    "appName": "IRI",
    "appVersion": "1.4.1.2",
    "jreAvailableProcessors": 4,
    "jreFreeMemory": 408453744,
    "jreVersion": "1.8.0_151",
    "jreMaxMemory": 3817865216,
    "jreTotalMemory": 2592604160,
    "latestMilestone": "9NUUINVENCYBQXDIUTDDOKMFVDLKWMNJLUNBTKIQRKVXPH9CROKOKIPAHIQIDNARNSWZUMQDOTXTA9999",
    "latestMilestoneIndex": 305511,
    "latestSolidSubtangleMilestone": "QYKENTHDNINZUPUM99FCDYNREQSMPPRLDYJEBWZXWQFD9UHPTJPFTPRSKKKCJOZUKGXKUVCQYAQK99999",
    "latestSolidSubtangleMilestoneIndex": 305408,
    "neighbors": 5,
    "packetsQueueSize": 0,
    "time": 1513547084402,
    "tips": 3297,
    "transactionsToRequest": 3,
    "duration": 12
}

root@...:~# iptables -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

root@....:~# ./nelson.sh
Unhandled Rejection at: Promise Promise {
  <rejected> Error: Invalid Response: Error: read ECONNRESET
    at _errnoException (util.js:999:13)
    at TCP.onread (net.js:629:25)
    at Object.invalidResponse (/usr/lib/node_modules/nelson.cli/node_modules/iota.lib.js/lib/errors/requestErrors.js:5:12)
    at makeRequest.prepareResult (/usr/lib/node_modules/nelson.cli/node_modules/iota.lib.js/lib/utils/makeRequest.js:160:24)
    at exports.XMLHttpRequest.request.onreadystatechange (/usr/lib/node_modules/nelson.cli/node_modules/iota.lib.js/lib/utils/makeRequest.js:62:25)
    at exports.XMLHttpRequest.dispatchEvent (/usr/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:591:25)
    at setState (/usr/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:610:14)
    at exports.XMLHttpRequest.handleError (/usr/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:532:5)
    at ClientRequest.errorHandler (/usr/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:459:14)
    at ClientRequest.emit (events.js:159:13)
    at Socket.socketErrorListener (_http_client.js:389:9)
    at Socket.emit (events.js:159:13) } reason: Error: Invalid Response: Error: read ECONNRESET
    at _errnoException (util.js:999:13)
    at TCP.onread (net.js:629:25)
    at Object.invalidResponse (/usr/lib/node_modules/nelson.cli/node_modules/iota.lib.js/lib/errors/requestErrors.js:5:12)
    at makeRequest.prepareResult (/usr/lib/node_modules/nelson.cli/node_modules/iota.lib.js/lib/utils/makeRequest.js:160:24)
    at exports.XMLHttpRequest.request.onreadystatechange (/usr/lib/node_modules/nelson.cli/node_modules/iota.lib.js/lib/utils/makeRequest.js:62:25)
    at exports.XMLHttpRequest.dispatchEvent (/usr/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:591:25)
    at setState (/usr/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:610:14)
    at exports.XMLHttpRequest.handleError (/usr/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:532:5)
    at ClientRequest.errorHandler (/usr/lib/node_modules/nelson.cli/node_modules/xmlhttprequest/lib/XMLHttpRequest.js:459:14)
    at ClientRequest.emit (events.js:159:13)
    at Socket.socketErrorListener (_http_client.js:389:9)
    at Socket.emit (events.js:159:13)
```

Maybe my exec is wrong (without --gui):

```
nelson --neighbors "mainnet.deviota.com/16600 mainnet2.deviota.com/16600 mainnet3.deviota.com/16600 iotairi.tt-tec.net/16600"
```

If I try with gui it just pops up for less than a second.

Why should I use docker? I have Ubuntu 17, nodeJS, Java and IRI on it. What else does it need to run?

**Comment by H3n-42** [comment]
@ortreum please hit me up in slack (@H3n) and or create a new issue, since this is not the right place (or do you run docker?)

**Comment by romansemko** [comment]
@ortreum , make sure your IRI has API listening locally on the default port: 14265. Otherwise you will need to provide Nelson with a correct port.

**Comment by romansemko** [comment]
@rosstimson , I will fix that. Makes sense!

**Comment by romansemko** [comment]
Fixed with 0.2.1
